app/routes.py

```diff
@@ -1 +1,190 @@
+    task = Task.query.get(task_id)
+    if task == None:
+        abort(404) 
+    else:
+        task.completed_at = time_stamp
+    db.session.commit()
+    is_complete = task.completed_at != None
+    
+    try:
+        header= {"Authorization"}
+        post_body = requests.post(header)
+    except TypeError:
+        pass
+    return make_response({"task": {"id": task.task_id, "title": task.title, "description": task.description, "is_complete": is_complete}}, 200)
+
+@tasks_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def task_not_complete(task_id):
+    task = Task.query.get(task_id)
+    if task == None:
+        abort(404) 
+    else:
+        task.completed_at = None
+    db.session.commit()
+    is_complete = task.completed_at != None
+    return make_response({"task": {"id": task.task_id, "title": task.title, "description": task.description, "is_complete": is_complete}}, 200)  
+
+@goals_bp.route("", methods=["POST"])
+def post_goal():
+    request_body = request.get_json()
+    if 'title' not in request_body :
+        return make_response({"details": "Invalid data"}, 400)
+    else:
+        new_goal = Goal(title=request_body["title"])
+
+        db.session.add(new_goal)
+        db.session.commit()
+        return make_response({"goal": {"id": new_goal.goal_id, "title": new_goal.title}}, 201)
+
+@goals_bp.route("", methods=["GET"])
+def all_goals():
+    title_query = request.args.get('title')
+    order_by_query = request.args.get('sort')
+    if title_query:
+        goals = Goal.query.filter_by(title = title_query)
+    elif order_by_query == 'asc':
+        goals = Goal.query.order_by(Goal.title).all()
+    elif order_by_query == 'desc':   
+        goals = Goal.query.order_by(desc(Goal.title)).all()
+    else:
+        goals = Goal.query.order_by(Goal.title).all()
+    
+    goal_response = []
+    for goal in goals:
+        goal_response.append({'id': goal.goal_id, 'title': goal.title})
+    return jsonify(goal_response), 200    
+
+@goals_bp.route("/<goal_id>", methods=["GET", "PUT", "DELETE"])
+def one_goal(goal_id):
+    goal = Goal.query.get(goal_id) 
+    if goal == None:
+        abort(404) 
+    if request.method == "GET":
+        return make_response({"goal": {"id": goal.goal_id, "title": goal.title}}, 200)
+    elif request.method == "PUT":
+        form_data = request.get_json()
+        goal.title = form_data["title"]
+        db.session.commit()
+
+        return make_response({"goal": {"id": goal.goal_id, "title": goal.title}}, 200)
+    elif request.method == "DELETE":
+        db.session.delete(goal)
+        db.session.commit()
+        return make_response({'details': 
+        f'Goal {goal.goal_id} "{goal.title}" successfully deleted'}, 200)
+
+@goals_bp.route("/<goal_id>/tasks", methods=["POST", "GET"]) 
+def task_and_goal(goal_id):
+    request_body = request.get_json()
+    tasks = Task.query.all()
+    goal = Goal.query.get(goal_id) 
+    if goal == None:
+        abort(404) 
```

Comment:
Here is another place we can create a helper function. Let's validate the parameter passed in through the route to check if it's the right data type, exists in our db, etc. We should do this for bouth our models.

Here's a class resource we made about this from [Flasky](https://github.com/Ada-C18/flasky/blob/tigers/app/routes/cat_routes.py)